src/dmd/dtemplate.d

```diff
@@ -2422,6 +2422,17 @@ void functionResolve(Match* m, Dsymbol dstart, Loc loc, Scope* sc, Objects* tiar
             else
                 return 0;   // MATCH.nomatch
         }
+        auto dt = fd.isDtorDeclaration();
+        if (dt)
+        {
+            auto dtmod = dt.type.toTypeFunction();
+            auto shared_dtor = dtmod.mod & MODFlags.shared_;
+            auto shared_this = tthis_fd.mod & MODFlags.shared_;
+            if (shared_dtor && !shared_this)
+                tthis_fd = dtmod;
+            else if (shared_this && !shared_dtor)
+                tf.mod = tthis_fd.mod;
```

Comment:
What's about the case of `shared_this && shared_dtor`?

Reply:
If either the struct/class declaration is shared or the destructor is shared, it makes sure that when the matching is done, both are considered shared.

Reply:
(I asked because such motivation belongs into the source code - it will help people in the future and also the review of this PR)